web-common/src/features/dashboards/proto-state/dashboard-url-state.ts

```diff
@@ -0,0 +1,87 @@
+import { goto } from "$app/navigation";
+import { page } from "$app/stores";
+import {
+  metricsExplorerStore,
+  useDashboardStore,
+} from "@rilldata/web-common/features/dashboards/dashboard-stores";
+import type { V1MetricsView } from "@rilldata/web-common/runtime-client";
+import type { CreateQueryResult } from "@tanstack/svelte-query";
+import { derived, get, Readable } from "svelte/store";
+
+export type DashboardUrlState = {
+  proto: string;
+  defaultProto: string;
+  urlProto: string;
+};
+export type DashboardUrlStore = Readable<DashboardUrlState>;
+
+/**
+ * Creates a derived store that has the current proto and default proto of a dashboard along with the proto in the url.
+ */
+export function useDashboardUrlState(
+  metricViewName: string
+): DashboardUrlStore {
+  return derived(
+    [
+      useDashboardProto(metricViewName),
+      useDashboardDefaultProto(metricViewName),
+      page,
+    ],
+    ([proto, defaultProto, page]) => {
+      let urlProto = page.url.searchParams.get("state");
+      if (urlProto) urlProto = decodeURIComponent(urlProto);
+      else urlProto = defaultProto;
+
+      return {
+        proto,
+        defaultProto,
+        urlProto,
+      };
+    }
+  );
+}
+
+/**
+ * Code that looks at dashboard state and url state and decides which one to sync with.
+ */
+export function useDashboardUrlSync(
+  metricViewName: string,
+  metaQuery: CreateQueryResult<V1MetricsView>
+) {
+  const dashboardUrlState = useDashboardUrlState(metricViewName);
+  let desiredProto: string;
```

Comment:
It's not really the `desiredProto` is it? Its more like the `lastKnownProto`?
Because you end up overwriting it any time its not the same as either the local state or the URL state.

With the current name, it makes it hard for me to understand what happens on loading a dash for first time, where desiredProto === undefined